src/app/payments/send-lightning.ts

```diff
@@ -0,0 +1,653 @@
+
+  return {
+    senderWallet,
+    paymentFlow,
+    decodedInvoice,
+  }
+}
+
+const newCheckAndVerifyTwoFA = async ({
+  amount,
+  twoFAToken,
+  twoFASecret,
+  wallet,
+  priceRatio,
+}: {
+  amount: UsdPaymentAmount
+  twoFAToken: TwoFAToken | null
+  twoFASecret: TwoFASecret
+  wallet: Wallet
+  priceRatio: PriceRatio
+}): Promise<true | ApplicationError> => {
+  const twoFALimitCheck = await newCheckTwoFALimits({
+    amount,
+    wallet,
+    priceRatio,
+  })
+  if (!(twoFALimitCheck instanceof Error)) return true
+
+  if (!twoFAToken) return new TwoFANewCodeNeededError()
+
+  const validTwoFA = TwoFA().verify({
+    secret: twoFASecret,
+    token: twoFAToken,
+  })
+  if (validTwoFA instanceof Error) return validTwoFA
+
+  return true
+}
+
+const executePaymentViaIntraledger = async ({
+  paymentFlow,
+  senderWallet,
+  logger,
+  senderUsername,
+  memo,
+}: {
+  paymentFlow: PaymentFlow<WalletCurrency, WalletCurrency>
+  senderWallet: Wallet
+  logger: Logger
+  senderUsername: Username | undefined
+  memo: string | null
+}): Promise<PaymentSendStatus | ApplicationError> => {
+  const priceRatio = PriceRatio({
+    usd: paymentFlow.usdPaymentAmount,
+    btc: paymentFlow.btcPaymentAmount,
+  })
+  if (priceRatio instanceof Error) return priceRatio
+
+  const limitCheck = await newCheckIntraledgerLimits({
+    amount: paymentFlow.usdPaymentAmount,
+    wallet: senderWallet,
+    priceRatio,
+  })
+  if (limitCheck instanceof Error) return limitCheck
+
+  const { paymentHash } = paymentFlow
+
+  const {
+    recipientWalletId,
+    recipientPubkey,
+    recipientWalletCurrency,
+    recipientUsername,
+  } = paymentFlow.recipientDetails()
+  if (!(recipientWalletId && recipientWalletCurrency && recipientPubkey)) {
+    return new InvalidLightningPaymentFlowBuilderStateError(
+      "Expected recipient details missing",
+    )
+  }
+  const recipientWallet = await WalletsRepository().findById(recipientWalletId)
+  if (recipientWallet instanceof Error) return recipientWallet
+
+  return LockService().lockWalletId(
+    { walletId: senderWallet.id, logger },
+    async (lock) => {
+      const balance = await LedgerService().getWalletBalanceAmount({
+        walletId: senderWallet.id,
+        walletCurrency: senderWallet.currency,
+      })
+      if (balance instanceof Error) return balance
+
+      const paymentAmount = paymentFlow.paymentAmountInSenderWalletCurrency()
+
+      if (balance.amount < paymentAmount.amount) {
+        const unitForMsg = senderWallet.currency === WalletCurrency.Btc ? "sats" : "cents"
+        return new InsufficientBalanceError(
+          `Payment amount '${paymentAmount.amount}' ${unitForMsg} exceeds balance '${balance.amount}'`,
+        )
+      }
+
+      const journal = await LockService().extendLock({ logger, lock }, async () => {
+        const lnIntraLedgerMetadata = LedgerFacade.LnIntraledgerLedgerMetadata({
+          // FIXME: display currency
+          amountDisplayCurrency: Number(
```

Comment:
How are we thinking to calculate this?

Right now I'm just passing the USD-equivalent amount straight through (can be from either dealer or price service). Should we explicitly add another step here to convert from the USD-amount to the display currency amount? It'll be a no-op for USD display currency, but will be applicable for CRC (cc @dolcalmi)

Reply:
I think it needs to be an explicit step that no-ops internally for usd -> usd cases (as suggested)

Reply:
in this moment there is no need but we can have an additional step to get display currency amount (probably  is better to pass 'base' and 'quote' to avoid to have a different step for each base currency)

also, you can use `displayCurrency.code` config to pass the quote (in the display currency PR will be a config from the user entity.. like language)